### ISSUE TITLE
Validation dataloader(s) + function

### ISSUE THREAD

**Issue by alecgunny** [ML4GW/BBHNet#26]
Some ideas on how validation might get implemented in the context of BBHNet

- Fixed dataset of background, glitches, injections
    - For injections, this includes the ifo response. Everything should be deterministic
    - Do we inject on the same data we use as background?
    - Do we use timeslides?
    - Do we use glitches from other observation periods and insert them like we do in the training set, or do we monitor our performance on a handful of "real" glitches from the validation period?
    - What does the class balance look like?
- Everything loaded into GPU memory and preprocessed up front
- Validation metrics
    - Always measure loss for comparison with training
    - Measure the AUC at various time slices using ROCs outlined in #5 ? Or FAR at a fixed TPR for various time slices?
    - Which metric do we use for model selection?
    - Monitor glitch performance with special metric?
        - Is there anyway to make FAR in units of yr^{-1} meaningful here? If we use inserted glitches and timeslides, the effective "analysis period" and glitch probability as a function of time are both made basically arbitrary. Does this metric still have meaning?

**Comment by eric-moreno** [comment]
I guess I'm also considering the reach of this project beyond the reach of just BBH, which is why I occasionally reference exotic waveforms. Here are my thoughts on your questions. 

Do we inject on the same data we use as background?
    > This makes sense to me. From whatever LIGO-run our background comes from, we should also use for validation. In terms of our constantly retraining models, I think that it would make sense to pull validation from the same (n)-th background period or the (n+1)th background period. Not much use in trying to generalize to a different period other than these two time periods right?
Do we use timeslides?
    >  Could make sense to save some timeslides for training/validation/test periods. 
Do we use glitches from other observation periods and insert them like we do in the training set, or do we monitor our performance on a handful of "real" glitches from the validation period?
    > I guess it depends on how generalizable we are trying to make this project (even for training set as well). For example, using a wide variety of glitches will increase our generalizability, but will also make us less sensitive to the more common glitches. Only using "real" glitches from the training/validation period will get us better sensitivity but decrease generalizability to the rare glitches. Maybe there is a good middle ground? 
What does the class balance look like?
    > I think this ties back to my previous point. Not sure which "classes" we speak of though. If anything for validation, we might want to consider glitch and signal classes outside the typical regime  since most purely classical BBH will be detected by matched filters.
    
 Measure the AUC at various time slices using ROCs outlined in https://github.com/ML4GW/BBHNet/issues/5 ? Or FAR at a fixed TPR for various time slices?
     > I think we fix FAR = 1 yr^{-1}. AUC is less important in our case since they likely wont care about high FARs (which contribute significantly to AUC integral. 
Which metric do we use for model selection?
    > There could be multiple metrics for multiple purposes. FAR = 1 yr^{-1} will give us the best performance overall. But there can be an additional metric with exotic validation. 
Monitor glitch performance with special metric?
    >  As I said above, could be an "in-observation-run" glitch performance metric and an "out-of-observation-run" glitch performance. 
Is there anyway to make FAR in units of yr^{-1} meaningful here? If we use inserted glitches and timeslides, the effective "analysis period" and glitch probability as a function of time are both made basically arbitrary. Does this metric still have meaning?
    > In order to make FAR meaningful, glitches would have to be inserted with some probability that mimics the rate of actual glitches in the detector. I would think that there is some distribution that we can probably pull from the LIGO runs?
